hudi-common/src/main/java/org/apache/hudi/common/table/timeline/CompletionTimeQueryView.java

```diff
@@ -175,42 +187,111 @@ public Option<String> getCompletionTime(String startTime) {
+   * @param rangeStart   The query range start completion time.
+   * @param rangeEnd     The query range end completion time.
+   * @param rangeType    The range type.
    *
-   * @return The instant time set.
+   * @return The sorted instant time list.
    */
-  public Set<String> getStartTimeSet(String startCompletionTime, String endCompletionTime) {
+  public List<String> getStartTimes(
+      HoodieTimeline readTimeline,
+      Option<String> rangeStart,
+      Option<String> rangeEnd,
+      InstantRange.RangeType rangeType) {
     // assumes any instant/transaction lasts at most 1 day to optimize the query efficiency.
-    return getStartTimeSet(startCompletionTime, endCompletionTime, s -> HoodieInstantTimeGenerator.instantTimeMinusMillis(s, MILLI_SECONDS_IN_ONE_DAY));
+    return getStartTimes(readTimeline, rangeStart, rangeEnd, rangeType, s -> HoodieInstantTimeGenerator.instantTimeMinusMillis(s, MILLI_SECONDS_IN_ONE_DAY));
   }
 
   /**
    * Queries the instant start time with given completion time range.
    *
-   * @param startCompletionTime   The start completion time.
-   * @param endCompletionTime     The end completion time.
-   * @param earliestStartTimeFunc The function to generate the earliest start time boundary
-   *                              with the minimum completion time {@code startCompletionTime}.
+   * @param rangeStart              The query range start completion time.
+   * @param rangeEnd                The query range end completion time.
+   * @param earliestInstantTimeFunc The function to generate the earliest start time boundary
+   *                                with the minimum completion time.
    *
-   * @return The instant time set.
+   * @return The sorted instant time list.
    */
-  public Set<String> getStartTimeSet(String startCompletionTime, String endCompletionTime, Function<String, String> earliestStartTimeFunc) {
-    String startInstant = earliestStartTimeFunc.apply(startCompletionTime);
+  @VisibleForTesting
+  public List<String> getStartTimes(
+      String rangeStart,
+      String rangeEnd,
+      Function<String, String> earliestInstantTimeFunc) {
+    return getStartTimes(metaClient.getCommitsTimeline().filterCompletedInstants(), Option.ofNullable(rangeStart), Option.ofNullable(rangeEnd),
+        InstantRange.RangeType.CLOSE_CLOSE, earliestInstantTimeFunc);
+  }
+
+  /**
+   * Queries the instant start time with given completion time range.
+   *
+   * @param readTimeline            The read timeline.
+   * @param rangeStart              The query range start completion time.
+   * @param rangeEnd                The query range end completion time.
+   * @param rangeType               The range type.
+   * @param earliestInstantTimeFunc The function to generate the earliest start time boundary
+   *                                with the minimum completion time.
+   *
+   * @return The sorted instant time list.
+   */
+  public List<String> getStartTimes(
```

Comment:
Is it possible for us to return the completion time directly?

Reply:
The completion time is user specified, the main purpose of this clazz is transfering the ocmpletion time range into instant time range.

Reply:
This is confusing with `startTime` and `endTime` in `IncrementalQueryAnalyzer`. Let's assume each instant has start timestamp and completion timestamp. Then, we can rename this method to `getInstantStartTimestamps`?